File: src/axonize/integrity.py
from __future__ import annotations

from axonize.models import Edge, Node, RelationType


def validate_graph(nodes: list[Node], edges: list[Edge]) -> list[str]:
    errors: list[str] = []
    seen: set[str] = set()
    node_ids: set[str] = set()

    for node in nodes:
        if node.id in seen:
            errors.append(f"duplicate_node_id:{node.id}")
        seen.add(node.id)
        node_ids.add(node.id)

    for edge in edges:
        if edge.source not in node_ids:
            errors.append(f"orphan_edge_source:{edge.source}")
        if edge.target not in node_ids:
            errors.append(f"orphan_edge_target:{edge.target}")
        if not RelationType.has_value(edge.type.value):
            errors.append(f"invalid_relation_type:{edge.type.value}")

    return errors


def repair_graph(nodes: list[Node], edges: list[Edge]) -> tuple[list[Node], list[Edge], list[str]]:
    warnings: list[str] = []

    unique_nodes: dict[str, Node] = {}
    for node in nodes:
        if node.id in unique_nodes:
            warnings.append(f"repair:removed_duplicate_node:{node.id}")
            continue
        unique_nodes[node.id] = node

    node_ids = set(unique_nodes)

    unique_edges: dict[tuple[str, str, str, str], Edge] = {}
    for edge in edges:
        if edge.source not in node_ids or edge.target not in node_ids:
            warnings.append(f"repair:removed_orphan_edge:{edge.source}->{edge.target}")
            continue
        if not RelationType.has_value(edge.type.value):
            warnings.append(f"repair:removed_invalid_relation:{edge.type.value}")
            continue
        key = (edge.source, edge.target, edge.type.value, edge.origin.value)
        existing = unique_edges.get(key)
        if existing is None or edge.weight > existing.weight or edge.confidence > existing.confidence:
            unique_edges[key] = edge

    repaired_nodes = list(unique_nodes.values())
    repaired_nodes.sort(key=lambda node: (node.file, node.id))
    repaired_edges = list(unique_edges.values())
    repaired_edges.sort(key=lambda edge: (edge.source, edge.type.value, edge.target, edge.origin.value))
    return repaired_nodes, repaired_edges, warnings

File: desktop/main/semantic/integrity.ts
import { Edge, Node, isRelationType } from "./models";

export function validateGraph(nodes: readonly Node[], edges: readonly Edge[]): string[] {
  const errors: string[] = [];
  const nodeIds = new Set<string>();

  for (const node of nodes) {
    if (nodeIds.has(node.id)) {
      errors.push(`duplicate_node_id:${node.id}`);
    }
    nodeIds.add(node.id);
  }

  for (const edge of edges) {
    if (!nodeIds.has(edge.source)) {
      errors.push(`orphan_edge_source:${edge.source}`);
    }
    if (!nodeIds.has(edge.target)) {
      errors.push(`orphan_edge_target:${edge.target}`);
    }
    if (!isRelationType(edge.type)) {
      errors.push(`invalid_relation_type:${edge.type}`);
    }
  }

  return errors;
}

export function repairGraph(nodes: readonly Node[], edges: readonly Edge[]): { nodes: Node[]; edges: Edge[]; warnings: string[] } {
  const warnings: string[] = [];
  const uniqueNodes = new Map<string, Node>();

  for (const node of nodes) {
    if (uniqueNodes.has(node.id)) {
      warnings.push(`repair:removed_duplicate_node:${node.id}`);
      continue;
    }
    uniqueNodes.set(node.id, node);
  }

  const nodeIds = new Set(uniqueNodes.keys());
  const uniqueEdges = new Map<string, Edge>();

  for (const edge of edges) {
    if (!nodeIds.has(edge.source) || !nodeIds.has(edge.target)) {
      warnings.push(`repair:removed_orphan_edge:${edge.source}->${edge.target}`);
      continue;
    }
    if (!isRelationType(edge.type)) {
      warnings.push(`repair:removed_invalid_relation:${edge.type}`);
      continue;
    }
    const key = `${edge.source}::${edge.target}::${edge.type}::${edge.origin}`;
    const existing = uniqueEdges.get(key);
    if (!existing || edge.weight > existing.weight || edge.confidence > existing.confidence) {
      uniqueEdges.set(key, edge);
    }
  }

  const repairedNodes = [...uniqueNodes.values()].sort((a, b) => `${a.file}|${a.id}`.localeCompare(`${b.file}|${b.id}`));
  const repairedEdges = [...uniqueEdges.values()].sort((a, b) =>
    `${a.source}|${a.type}|${a.target}|${a.origin}`.localeCompare(`${b.source}|${b.type}|${b.target}|${b.origin}`)
  );
  return { nodes: repairedNodes, edges: repairedEdges, warnings };
}
